crates/rpc/ipc/src/server/connection.rs

```diff
@@ -145,28 +146,29 @@ impl<T, S> Future for IpcConnDriver<T, S, S::Future>
 where
     S: Service<String, Response = Option<String>> + Send + 'static,
     S::Error: Into<Box<dyn std::error::Error + Send + Sync>>,
-    S::Future: Send,
+    S::Future: Send + Unpin,
     T: AsyncRead + AsyncWrite + Unpin + Send + 'static,
 {
     type Output = ();
 
     fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
         let mut this = self.project();
 
-        loop {
-            // process calls
-            if !this.pending_calls.is_empty() {
-                while let Poll::Ready(Some(res)) = this.pending_calls.as_mut().poll_next(cx) {
-                    let item = match res {
-                        Ok(Some(resp)) => resp,
-                        Ok(None) => continue,
-                        Err(err) => err.into().to_string(),
-                    };
-                    this.items.push_back(item);
-                }
+        // items are also pushed from external
+        // this will act as a manual yield point to reduce latencies of the polling future that may
+        // submit items from an additional source (subscription)
+        let mut budget = 5;
+
+        // ensure we still have enough budget for another iteration
+        'outer: loop {
+            budget -= 1;
+            if budget == 0 {
+                // make sure we're woken up again
+                cx.waker().wake_by_ref();
+                return Poll::Pending
             }
```

Comment:
not quite clear to me - why do we need to return early? is it because the task is blocking and you want to yield back control to the scheduler?

Reply:
yes, this ensures this loops doesn't run for too long, since this is polled concurrently with something that also receives responses (from subscriptions)